src/mock_network.zig

```diff
@@ -0,0 +1,229 @@
+const std = @import("std");
+const assert = std.debug.assert;
+const log = std.log.scoped(.mock_network);
+
+pub const MockNetworkOptions = struct {
+    /// Mean for the exponential distribution used to calculate forward delay.
+    forward_delay_mean: u64,
+
+    /// Mean for the exponential distribution used to calculate reverse delay.
+    reverse_delay_mean: u64,
+    min_forward_delay: u64,
+    min_reverse_delay: u64,
+    packet_loss_probability: u8,
+    packet_replay_probability: u8,
+    prng_seed: u64,
+    node_count: u8,
+
+    /// The maximum number of in-flight packets a path can have before packets are randomly dropped.
+    path_maximum_capacity: u8,
+
+    /// Mean for the exponential distribution used to calculate how long a path is clogged for.
+    path_clog_duration_mean: u64,
+    path_clog_probability: u8,
+};
+
+/// A fully connected network of nodes used for testing. Simulates the fault model:
+/// Packets may be dropped.
+/// Packets may be delayed.
+/// Packets may be replayed.
+pub const PacketStatistics = enum(u8) {
+    dropped_due_to_congestion,
+    dropped,
+    forward_replay,
+    reverse_replay,
+};
+pub fn MockNetwork(comptime Packet: type) type {
+    return struct {
+        const Self = @This();
+        const Direction = enum(u8) {
+            forward,
+            reverse,
+        };
+        const Data = struct {
+            expiry: u64,
+            callback: fn (
+                packet: Packet,
+                to: u8,
+                from: u8,
+            ) void,
+            packet: Packet,
+            direction: Direction,
```

Comment:
Why do we need this state? We should already have a separate queue for each direction of a connection on our paths slice, one at `self.path(a, b)` and the other at `self.path(b, a)`.